### PR TITLE
handle sparse gradients

### PR DESCRIPTION
**Pull request by justheuristic**
This small fix allows using gradient checkpointing even if some gradients are sparse.

For instance, if one uses Embedding layer and requests derivatives w.r.t. embedding weights, tensorflow.gradients would produce gradient as IndexedSlices instead of tensor.

Doing the same with gradient checkpointing results in an error on line
```
d_xs[j] += d_xs_new[j]
```
Here, both d_xs[j] += d_xs_new[j] would be tf.IndexedSlices that cannot be added with '+'.
A cheap solution is to manually cast everything to dense gradients if addition is required.

Given enough time, one could also devise a way to add indexed slices sparsely. Alas, i am not given enough time.

### CODE REVIEW

**Comment by yaroslavvb** [comment]
Hm I don't seem to have merge rights to this repo, @TimSalimans do you know who could approve the merge?
